src/lib/gadgets/bitwise.ts

```diff
@@ -10,7 +10,44 @@ import {
   divideWithRemainder,
 } from './common.js';
 
-export { xor, and, rotate };
+export { xor, not, and, rotate };
+
+function not(a: Field, length: number) {
+  // check that input length is positive
+  assert(length > 0, `Input length needs to be positive values.`);
+
+  // Check that length does not exceed maximum field size in bits
+  assert(
+    length <= Field.sizeInBits(),
+    `Length ${length} exceeds maximum of ${Field.sizeInBits()} bits.`
+  );
+
+  // obtain pad length until the length is a multiple of 16 for n-bit length lookup table
+  let padLength = Math.ceil(length / 16) * 16;
+
+  // handle constant case
+  if (a.isConstant()) {
+    let max = 1n << BigInt(padLength);
+    assert(
+      a.toBigInt() < max,
+      `${a.toBigInt()} does not fit into ${padLength} bits`
+    );
+    return new Field(Fp.not(a.toBigInt(), length));
+  }
+
+  // create a bitmask with all ones
+  let allOnesF = new Field(2n ** BigInt(length) - 1n);
+
+  let allOnes = Provable.witness(Field, () => {
+    return allOnesF;
+  });
+
+  allOnesF.assertEquals(allOnes);
```

Comment:
I was wondering why we can't just pass in `allOnesF` to `xor()`. I assume this was done because otherwise when passing the constant to the xor16 gate, snarky would add a generic gate to resolve the constant, and that would break the xor chain.

I think it's actually a major gotcha that should be handled _in `xor()`_. I.e. `xor()` should do something like this when one of its inputs is constant. (Assuming it's really a problem)

@Trivo25 you wrote xor, do you agree?

Reply:
I am not sure, to be honest

I checked that if one of the inputs is a constant that the chain doesnt break, and it seems to work

```ts
let cs = Provable.constraintSystem(() => {
  let variable = Provable.witness(Field, () => Field(5));
  let constant = Field(4);
  Gadgets.xor(variable, constant, 128);
});

console.log(cs);
```

```
  gates: [
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Xor16', wires: [Array], coeffs: [] },
    { type: 'Zero', wires: [Array], coeffs: [] },
    { type: 'Generic', wires: [Array], coeffs: [Array] }
  ],
```

Reply:
perhaps @querolita can explain why we do this check? What are your thoughts?

Reply:
Right - because the constant resolving already happens _before_ the first xor16 gate, and after that it's only witnesses, no constants

Reply:
Do you think there is still an issue? If so, I think I am having trouble understanding :D

Reply:
Seems like there's no issue with just passing a constant directly 👍🏻

Reply:
ahh amazing :D